### PR TITLE
Registry API On-behalf-of calls auth check fix

### PR DESCRIPTION
**Pull request by varunpuranik**
Issue: The Registry APIs exposed by the EdgeHub are used to CRUD module identities. 
For nested scenarios, special on-behalf-of APIs are supported, which allows the parent EdgeHub to CRUD module identities on behalf of the child. 
Imagine a topology like l5->l4->l3. If a module is added to l3, then the EdgeHub on l4 and l5 need to make the call to create the module on l3. For this they use the "on-behalf-of" api. Same applies to list/delete as well. 
But currently the code does not check if the caller is making the call on behalf of a descendant. This allowed, for example, l4, to CRUD modules on behalf of l5!

Fix:
Get the auth chain from the current EdgeHub to the target, and make sure the caller is on that auth chain. For example, then l4 makes a call to l5, to create a module on l3, then the auth chain from l3 to l5 is "l3;l4;l5". So l5 can validate that l4 is in the auth chain. 
Now if l4 makes a call to create a module on l5, then the auth chain from l5 to l5 is  "l5". l4 is not in this auth chain, so validation fails.

### CODE REVIEW

File: edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.Http.Test/RegistryControllerTest.cs [hunk 1]
```diff
@@ -298,6 +298,38 @@ public async Task DeleteModuleOnBehalfOf_Success()
             Assert.Empty(actualResponseBytes);
         }
 
+        [Theory]
+        [InlineData("l4", "l3;l4", "l3", "l3;l4;l5", true, HttpStatusCode.OK)]
+        [InlineData("l4", "l2;l3;l4", "l2", "l2;l3;l4;l5", true, HttpStatusCode.OK)]
+        [InlineData("l4", "l4", "l4", "l4;l5", true, HttpStatusCode.OK)]
+        [InlineData("l3", "l4", "l4", "l4;l5", true, HttpStatusCode.Unauthorized)]
+        [InlineData("l4", "l3;l4", "l3", null, true, HttpStatusCode.Unauthorized)]
+        [InlineData("l4", "l3;l4", "l3", "l3;l4;l5", false, HttpStatusCode.Unauthorized)]
```
Comment (richma-ms) [inline edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.Http.Test/RegistryControllerTest.cs:307]: Would be good to add the following cases for client-provided authchain:
repeated ID: "l4;l4;l4"
reversed ID: "l5;l4;l3"

**Review by richma-ms — DISMISSED** [review]

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/RegistryController.cs [hunk 2]
```diff
@@ -274,44 +279,24 @@ public class RegistryController : Controller
...
-                    string errorMessage = $"Target device Id does not match between auth chain ({targetDeviceId}) and request body ({requestData.Module.DeviceId}).";
-                    Events.BadRequest(nameof(this.CreateOrUpdateModuleOnBehalfOfAsync), errorMessage);
-                    await this.SendResponseAsync(HttpStatusCode.BadRequest, FormatErrorResponseMessage(errorMessage));
-                    return;
-                }
-
-                if (!await this.AuthenticateAsync(actorDeviceId, Option.Some(Constants.EdgeHubModuleId), Option.Some(requestData.AuthChain)))
+                IEdgeHub edgeHub = await this.edgeHubGetter;
+                IHttpRequestAuthenticator authenticator = await this.authenticatorGetter;
+                Try<string> targetAuthChainTry = await ValidateOnBehalfOfCall(actorDeviceId, requestData.AuthChain, nameof(this.CreateOrUpdateModuleAsync), this.HttpContext, edgeHub, authenticator);
```
Comment (ancaantochi) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/RegistryController.cs:284]: source should be: CreateOrUpdateModuleOnBehalfOfAsync

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/RegistryController.cs [hunk 3]
```diff
@@ -507,10 +455,53 @@ public class RegistryController : Controller
             }
         }
 
-        async Task<bool> AuthenticateAsync(string deviceId, Option<string> moduleId, Option<string> authChain)
+        internal static async Task<Try<string>> ValidateOnBehalfOfCall(
```
Comment (ancaantochi) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/RegistryController.cs:458]: nit: rename to AuthorizeOnBehalfOf

File: edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.Http.Test/RegistryControllerTest.cs [hunk 4]
```diff
@@ -298,6 +298,40 @@ public async Task DeleteModuleOnBehalfOf_Success()
             Assert.Empty(actualResponseBytes);
         }
 
+        [Theory]
+        [InlineData("l4", "l3;l4", "l3", "l3;l4;l5", true, HttpStatusCode.OK)]
+        [InlineData("l4", "l2;l3;l4", "l2", "l2;l3;l4;l5", true, HttpStatusCode.OK)]
+        [InlineData("l4", "l4", "l4", "l4;l5", true, HttpStatusCode.OK)]
+        [InlineData("l3", "l4", "l4", "l4;l5", true, HttpStatusCode.Unauthorized)]
```
Comment (ancaantochi) [inline edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.Http.Test/RegistryControllerTest.cs:305]: maybe add a case for  [InlineData("l3", "l3;l4", "l4", "l4;l5", true, HttpStatusCode.Unauthorized)]

**Review by ancaantochi — DISMISSED** [review]

File: edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.Http.Test/RegistryControllerTest.cs (hunk 4 above)
Comment (varunpuranik) [reply edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.Http.Test/RegistryControllerTest.cs:305]: No, that would not be a valid test. 
So the first 2 parameters are what we get from the caller, and we figure out 3 and 4. 
First is the caller Id, 2nd is the auth Chain. 
Target is the 1st element of the auth chain. I just pass it in here to not reuse the product code that gets the target from the chain. So if the chain is "l3;l4" then the target is "l3".

Comment (ancaantochi) [reply edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.Http.Test/RegistryControllerTest.cs:305]: ok, then [InlineData("l4", "l5;l4;l3", "l3", "l3;l4;l5", false, HttpStatusCode.Unauthorized)] should be modified to
[InlineData("l4", "l5;l4", "l5", "l5", false, HttpStatusCode.Unauthorized)] ?

**Review by ancaantochi — APPROVED** [review]

**Review by ancaantochi — APPROVED** [review]

**Pull request merged**